orttraining/orttraining/training_api/module.cc

```diff
@@ -2,14 +2,135 @@
+
+using namespace onnxruntime;
 
 namespace onnxruntime {
 namespace training {
 namespace api {
 
+Status Parameter::AllocateGrad(const std::string& gradient_name, const SessionState& sess_state) {
+  // assert param is allocated
+  ORT_ENFORCE(data_.IsAllocated(), "Parameter data should be allocated before allocating gradient.");
+  ORT_ENFORCE(requires_grad_, "Gradient should only be allocated for trainable parameters.");
+  gradient_name_ = gradient_name;
+  ORT_ENFORCE(utils::OrtValueLike(sess_state, data_, gradient_).IsOK(), "Failed to allocate gradient.");
+  return Status::OK();
+}
+
+Status Parameter::ResetGrad() {
+  if (!requires_grad_) {
+    return Status::OK();
+  }
+  // TODO: make use of lazy_reset_grad input instead. maybe rename method
+  Tensor* p_tensor = gradient_.GetMutable<Tensor>();
+  const auto& device = p_tensor->Location().device;
+  if (device.Type() == OrtDevice::CPU) {
+    memset(p_tensor->MutableDataRaw(), 0, p_tensor->SizeInBytes());
+  }
+#if defined(USE_CUDA) || defined(USE_ROCM)
+  else if (device.Type() == OrtDevice::GPU) {
+    ORT_NOT_IMPLEMENTED("Not implemented.");
+  }
+#endif
+  else {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Unknown device type ", device.Type(), " for param:", name_);
+  }
+  return Status::OK();
+}
+Module::Module(const std::string& train_model_path_or_bytes,
+               std::unordered_map<std::string, std::shared_ptr<Parameter>>& parameters,
+               const std::optional<std::string>& eval_model_path_or_bytes) : named_parameters_(parameters) {
+  // create value copy as same params are passed to Optimizer constructor
+
+  auto so = onnxruntime::SessionOptions();
+  std::unique_ptr<Environment> env;
+  ORT_THROW_IF_ERROR(Environment::Create(nullptr, env));
+  train_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+  ORT_THROW_IF_ERROR(train_sess_->Load(train_model_path_or_bytes));
+  ORT_THROW_IF_ERROR(train_sess_->Initialize());
+  if (eval_model_path_or_bytes.has_value()) {
+    // TODO:: do validation on eval inputs and outputs: eg order of user inputs, weights
+    eval_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+    ORT_THROW_IF_ERROR(eval_sess_->Load(eval_model_path_or_bytes.value()));
+    ORT_THROW_IF_ERROR(eval_sess_->Initialize());
+    utils::GetGraphInputOutputNames(eval_sess_, eval_input_names_, eval_output_names_);
+  }
+
+  utils::GetGraphInputOutputNames(train_sess_, train_input_names_, train_output_names_);
+  auto& train_sess_state = train_sess_->GetSessionState();
+  std::vector<std::string> param_input_names, grad_input_names, user_input_names, reset_grad_name;
+  std::string param_name;
+
+  for (const auto& input_name : train_input_names_) {
+    auto it = named_parameters_.find(input_name);
+    if (it != named_parameters_.end()) {
+      param_input_names.emplace_back(input_name);
+      weights_.push_back(it->second->Data());
+    } else if (input_name == utils::LAZY_RESET_GRAD_NAME) {
```

Comment:
you mean , the lazy reset name is a user inputs?

Reply:
it is a graph input handled by module internally. Basically what it will do is when we want to do a reset and then train step, in the InPlaceAccumulator node it will do a memcpy instead of an accumulate.